### PR TITLE
feat: lock

### PR DESCRIPTION
**Pull request by varas**
Lock to avoid simultaneous publishing clashing and corrupting packages metadata.

Current lock tries once or fails. Later on we could add a backoff and retry policy.

Expiration mechanism is based on TTL and flushed time into lockfile contents. We could use [s3 object expiration](https://aws.amazon.com/blogs/aws/amazon-s3-object-expiration/) but this solution was quicker.

Breakdown:
- [x] lock interface
- [x] lock for testing
- [x] lock using S3
- [x] integration with publisher
- [x] add aws secrets to repo
- [x] expiration (disabled)

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -10,6 +10,15 @@ inputs:
   aws_s3_bucket_name:
     description: Name of the S3 bucket
     required: true
+  aws_s3lock_bucket_name:
```
Comment (Noly) [inline action.yml:13]: there is a typo in `aws_s3lock_bucket_name` should be `aws_s3_lock_bucket_name`

File: publisher/lock/s3_lock.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		filePath: filepath,
+		tags:     fmt.Sprintf("department=product&product=%s&project=%s&owning_team=%s&environment=%s", tagProduct, tagProject, tagOwningTeam, tagEnv),
+		ttl:      defaultTTL,
+	}, nil
+}
+
+// Lock S3 has no compare-and-swap so this is no bulletproof solution, but should be good enough.
+func (l *S3) Lock() error {
+	if l.isBusyDeletingExpired() {
+		return LockBusyErr
```
Comment (Noly) [inline publisher/lock/s3_lock.go:84]: I believe to remember the naming convention for public error is to use `Err` as prefix, am I right?
I couldn't find the go doc about it thou :(

Comment (varas) [reply publisher/lock/s3_lock.go:84]: you are right, suffix for types, prefix for vars https://github.com/golang/go/wiki/Errors#naming

File: publisher/publisher_test.go [hunk 3]
```diff
@@ -247,6 +249,63 @@ func TestUploadArtifacts(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestUploadArtifacts_cantBeRunInParallel(t *testing.T) {
+	t.Skip("flaky on CI")
```
Comment (Noly) [inline publisher/publisher_test.go:253]: I appreciate the test but if don't provide any value maybe we should delete it.

I would prefer to have it working in CI thou. (IMHO)

**Review by Noly — APPROVED** [review]
Really clean code and easy to understand. Good job!!

Everything looks good to me 👍

File: publisher/publisher_test.go [hunk 4]
```diff
@@ -247,6 +249,63 @@ func TestUploadArtifacts(t *testing.T) {
...
+	var err1, err2 error
+	l := lock.NewInMemory()
+	go func() {
+		<-ready
+		err1 = uploadArtifacts(cfg, schema, l)
+		wg.Done()
+	}()
+	go func() {
+		<-ready
+		time.Sleep(1 * time.Millisecond)
```
Comment (cristianciutea) [inline publisher/publisher_test.go:295]: Could it be that this test is flaky because we wait to less here?

Comment (varas) [reply publisher/publisher_test.go:295]: yep, time based tests are never good, but code will need a revamp in order to test this without incurring into time delays

Comment (varas) [reply publisher/publisher_test.go:293]: restored

**Review by brushknight — APPROVED** [review]
LGTM

**Pull request merged**